pkg/proxy/iptables/proxier_test.go

```diff
@@ -1367,6 +1369,258 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+		}
+	}
+
+	// The sub-rules within an iptables rule are ANDed together, so the rule only
+	// matches if all of them match. So go through the subrules, and if any of them
+	// don't match, then fail.
+
+	// Match local/non-local
+	match = localRegex.FindStringSubmatch(rule)
+	if match != nil {
+		sourceIsLocal := sourceIP == nodeIP || sourceIP == "127.0.0.1"
+		destIsLocal := destIP == nodeIP || destIP == "127.0.0.1"
+
+		if match[1] == "" && match[2] == "src" && !sourceIsLocal {
+			return false
+		}
+		if match[1] == "!" && match[2] == "src" && sourceIsLocal {
+			return false
+		}
+		if match[1] == "" && match[2] == "dst" && !destIsLocal {
+			return false
+		}
+		if match[1] == "!" && match[2] == "dst" && destIsLocal {
+			return false
+		}
+	}
+
+	// Match destination IP/port
+	match = destIPRegex.FindStringSubmatch(rule)
+	if match != nil {
+		if match[1] == "!" && cidrContains(match[2], destIP) {
+			return false
+		} else if match[1] == "" && !cidrContains(match[2], destIP) {
+			return false
+		}
+	}
+	match = destPortRegex.FindStringSubmatch(rule)
+	if match != nil && match[1] != destPort {
+		return false
+	}
+
+	// Match source IP
+	match = sourceIPRegex.FindStringSubmatch(rule)
+	if match != nil {
+		if match[1] == "!" && cidrContains(match[2], sourceIP) {
+			return false
+		} else if match[1] == "" && !cidrContains(match[2], sourceIP) {
+			return false
+		}
+	}
+
+	// Match endpoint jump
+	if affinityRegex.MatchString(rule) {
+		// Ignore SessionAffinity rules
+		return false
+	}
+	match = probabilityRegex.FindStringSubmatch(rule)
+	if match != nil {
+		// There's an explicit probability on this line, so it only matches
+		// if it's the probability corresponding to "endpointProbability".
+		if match[1] != endpointProbability {
+			return false
+		}
+	} else if jumpToEndpointRegex.MatchString(rule) {
+		// This is the fallback endpoint jump, but we only want it to match if
+		// the caller requested the fallback endpoint (as opposed to having
+		// requested a previous endpoint that turned out to not exist).
+		if endpointProbability != "" {
+			return false
+		}
+	}
+
+	return true
+}
+
+// tracePacket determines what would happen to a packet with the given sourceIP, destIP,
+// and destPort, given the indicated iptables ruleData. nodeIP is the local node IP (for
+// rules matching "LOCAL"). endpoint indicates which endpoint to "randomly" select. (If
+// endpoint is 0 then the last endpoint jump will match. If endpoint is larger than the
+// actual number of endpoints then no endpoint will match.).
+//
+// The return values are: an array of matched rules, the final packet destination (or one
+// of the special strings "ACCEPT", "DROP", or "REJECT"), and whether the packet would be
+// masqueraded.
+func tracePacket(ruleData, sourceIP, destIP, destPort, nodeIP string, endpoint int) ([]string, string, bool) {
+	tables, err := parseIPTablesData(ruleData)
+	if err != nil {
+		panic("bad iptables ruleData: " + err.Error())
+	}
+
```

Comment:
I think that you should add the `checkIPTablesRuleJumps` as a safeguard so people that adds a test wrong don't create an endless loop

Reply:
`checkIPTablesRuleJumps` does not check that. It checks for:
- jumps to non-created chains (which would cause the e2e tests to fail too because `iptables-restore` won't let you do that, but this potentially catches it sooner)
- jumps to empty chains (which is a theoretically valid thing to do, and so would not cause `iptables-restore` failures, but which happens to be something we don't _actually_ do, so we know it's a bug if it happens)
- chains that never get jumped to (which again is theoretically valid, but not expected in our code)

I could modify `tracePacket` to keep track of the chains it passes through and notice if it ends up in a loop, but I'm not sure that's worthwhile? We never had any tests for this in the past and never had any problems... (Tim brought up the possibility of loops in the other PR, but that was a case where the code was (non-obviously) creating a SRV-to-XLB jump when we had a localDetector and an XLB-to-SRV jump when we didn't have a localDetector, so there would never have actually been any output where a loop was possible.)